stdlib/LinearAlgebra/src/symmetric.jl

```diff
@@ -578,9 +579,12 @@ end
 
 function dot(x::AbstractVector, A::RealHermSymComplexHerm, y::AbstractVector)
     require_one_based_indexing(x, y)
-    (length(x) == length(y) == size(A, 1)) || throw(DimensionMismatch())
+    n = length(x)
+    (n == length(y) == size(A, 1)) || throw(DimensionMismatch())
     data = A.data
-    r = zero(eltype(x)) * zero(eltype(A)) * zero(eltype(y))
+    r = dot(zero(eltype(x)), zero(eltype(A)), zero(eltype(y)))
+    iszero(n) && return r
+    isdiag(A) && return dot(x, Diagonal(A), y)
```

Comment:
It doesn't make sense to me to have a generic `isdiag` check in `dot` — isn't it the same complexity as the `dot` itself?

Reply:
You're absolutely right. I should've run benchmarks on the computer, and not just in my head. 🤦  If anything, then we should add what you're suggesting.